### PR TITLE
feat: EDGE sync

### PR DESCRIPTION
**Pull request by dmaretskyi**
### Motivation / Background

This Pull Request ...

### CODE REVIEW

File: packages/core/mesh/edge-client/src/client.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+const DEFAULT_TIMEOUT = 5_000;
+
+export type MessageListener = (message: Message) => void | Promise<void>;
+
+/**
+ *
+ */
+// TODO(dmaretskyi): Rename EdgeConnection
+export interface Messenger {
```
Comment (dmaretskyi) [inline packages/core/mesh/edge-client/src/client.ts:25]: rename EdgeClient

File: packages/core/echo/echo-db/src/edge/echo-edge-replicator.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    this.writable = new WritableStream<AutomergeMessage>({
+      write: async (message: AutomergeMessage, controller) => {
+        await this._sendMessage(message);
+      },
+    });
+  }
+
+  protected override async _open(ctx: Context): Promise<void> {
+    this._ctx.onDispose(
+      this._edgeConnection.addListener((msg: RouterMessage) => {
```
Comment (dmaretskyi) [inline packages/core/echo/echo-db/src/edge/echo-edge-replicator.ts:163]: Handle edge reconnects

File: packages/core/mesh/edge-client/src/client.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+   * Open connection to messaging service.
+   */
+  public async open(): Promise<boolean> {
+    invariant(!this._ws);
+    log.info('opening...', { info: this.info });
+    const ready = new Trigger<boolean>();
+
+    const url = new URL(`/ws/${this._identityKey.toHex()}/${this._deviceKey.toHex()}`, this._config.socketEndpoint);
+    this._ws = new WebSocket(url);
+    Object.assign<WebSocket, Partial<WebSocket>>(this._ws, {
```
Comment (dmaretskyi) [inline packages/core/mesh/edge-client/src/client.ts:95]: Must handle websocket reconnects

File: packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+  constructor({ messenger, spaceId }: EdgeFeedReplicatorParams) {
+    super();
+    this._messenger = messenger;
+    this._spaceId = spaceId;
+  }
+
+  protected override async _open(): Promise<void> {
+    this._ctx.onDispose(
+      this._messenger.addListener(async (message) => {
```
Comment (dmaretskyi) [inline packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts:49]: handle reconnects

File: packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    await this._sendMessage({
+      type: 'get-metadata',
+      feedKey: feed.key.toHex(),
+    });
+
+    Event.wrap(feed.core as any, 'append').on(this._connectionCtx, async () => {
+      await this._pushBlocksIfNeeded(feed);
+    });
+
+    // Event.wrap(feed.core as any, 'ready').on(this._connectionCtx, async () => {
```
Comment (dmaretskyi) [inline packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts:111]: remove

File: packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    hash: arrayToBuffer(node.hash),
+    size: node.size,
+  })),
+  signature: arrayToBuffer(block.signature),
+});
+
+// TODO(dmaretskyi): Protocol types.
+// TODO(dmaretskyi): Automerge core protocol vs websocket protocol.
+// TODO(dmaretskyi): Backpressure.
+// TODO(dmaretskyi): Worker share policy.
```
Comment (dmaretskyi) [inline packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts:256]: remove those

File: packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts [hunk 7]
```diff
@@ -0,0 +1,256 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { create } from '@bufbuild/protobuf';
```
Comment (dmaretskyi) [inline packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts:5]: rexport buf utils from protocols

File: packages/apps/testbench-app/dx.yml [hunk 8]
```diff
@@ -18,3 +19,6 @@ runtime:
       - urls: turn:kube.dxos.org:3478
         username: dxos
         credential: dxos
+    edge:
```
Comment (dmaretskyi) [inline packages/apps/testbench-app/dx.yml:22]: remove before merging

File: packages/apps/composer-app/dx.yml [hunk 9]
```diff
@@ -29,3 +30,6 @@ runtime:
     ipfs:
       server: https://api.ipfs.dxos.network/api/v0
       gateway: https://gateway.ipfs.dxos.network/ipfs
+    edge:
```
Comment (dmaretskyi) [inline packages/apps/composer-app/dx.yml:33]: remove before merging

File: packages/core/protocols/src/feed-replication.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
+//
```
Comment (dmaretskyi) [inline packages/core/protocols/src/feed-replication.ts:1]: Move automerge sync protocol to a module in @dxos/protocols as well

File: packages/core/protocols/src/proto/dxos/edge/messenger.proto [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+//
+message Message {
+  optional string timestamp = 1;
+
+  // Routing.
+  optional Peer source = 10;
+  repeated Peer target = 11;
+
+  // TODO(burdon): Routing info.
+  optional string service_id = 20;
```
Comment (dmaretskyi) [inline packages/core/protocols/src/proto/dxos/edge/messenger.proto:45]: define service ID constants

File: packages/core/echo/echo-db/src/edge/echo-edge-replicator.ts [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+      type: message.type,
+      senderId: message.senderId,
+      targetId: message.targetId,
+      documentId: message.documentId,
+    });
+    const encoded = cbor.encode(message);
+
+    await this._edgeConnection.send(
+      create(RouterMessageSchema, {
+        serviceId: `automerge-replicator:${this._spaceId}`,
```
Comment (dmaretskyi) [inline packages/core/echo/echo-db/src/edge/echo-edge-replicator.ts:250]: define constant

File: packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+    invariant(message.feedKey);
+    const payloadValue = bufferToArray(encodeCbor(message));
+
+    await this._messenger.send(
+      create(RouterMessageSchema, {
+        source: {
+          identityKey: this._messenger.identityKey.toHex(),
+          peerKey: this._messenger.deviceKey.toHex(),
+        },
+        serviceId: `hypercore-replicator:${this._spaceId}`,
```
Comment (dmaretskyi) [inline packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts:128]: define constant

File: packages/core/protocols/src/proto/dxos/edge/messenger.proto (hunk 11 above)
Comment (yarolegovich) [reply packages/core/protocols/src/proto/dxos/edge/messenger.proto:45]: can't use enums here cause service_ids are dynamic, defined prefixes in a code file instead

File: packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts (hunk 4 above)
Comment (yarolegovich) [reply packages/sdk/client-services/src/packlets/spaces/edge-feed-replicator.ts:50]: added TODOs for reconnects

**Pull request merged**